runtime/lua/vim/lsp/inlay_hint.lua

```diff
@@ -435,4 +436,212 @@ function M.enable(enable, filter)
+function M.apply_action(action)
+  local action_callback = action
+  if type(action) == 'string' then
+    action_callback = inlayhint_actions[action]
+    ---@cast action_callback -vim.lsp.inlay_hint.action.name
+  end
+  local bufnr = api.nvim_get_current_buf()
+  local winid = fn.bufwinid(bufnr)
+  local clients = vim.lsp.get_clients({ bufnr = bufnr, method = 'textDocument/inlayHint' })
+
+  local mode = fn.mode()
+
+  -- mark position, (1, 0) indexed, end-inclusive
+  ---@type {start: [integer, integer], end: [integer, integer]}
+  local range = {}
+
+  if mode == 'n' then
+    local cursor = api.nvim_win_get_cursor(winid)
+    range.start = cursor
+    range['end'] = cursor
+  else
+    local start_pos = fn.getpos('v')
+    local end_pos = fn.getpos('.')
+    if start_pos[2] > end_pos[2] or (start_pos[2] == end_pos[2] and start_pos[3] > end_pos[3]) then
+      ---@type [integer, integer, integer, integer]
+      start_pos, end_pos = end_pos, start_pos
+    end
+
+    range = {
+      start = { start_pos[2], start_pos[3] - 1 },
+      ['end'] = { end_pos[2], end_pos[3] - 2 },
+    }
+
+    if mode == 'V' or mode == 'Vs' then
+      range.start[2] = 0
+      range['end'][1] = range['end'][1] + 1
+      range['end'][2] = 0
+    end
+  end
+
+  ---@param idx? integer
+  ---@param client vim.lsp.Client
+  local function do_insert(idx, client)
+    if idx == nil then
+      return
+    end
+
+    local params =
+      util.make_given_range_params(range.start, range['end'], bufnr, client.offset_encoding)
+    local support_resolve = client:supports_method('inlayHint/resolve', bufnr)
+
+    client:request(
+      'textDocument/inlayHint',
+      params,
+      ---@param result lsp.InlayHint[]?
+      function(_, result, _, _)
+        if result ~= nil then
+          ---@type lsp.InlayHint[]
+          local hints = vim
+            .iter(result)
+            :filter(
+              ---@param hint lsp.InlayHint
+              function(hint)
+                local hint_pos = hint.position
+                if
+                  hint_pos.line < params.range.start.line
+                  or hint_pos.line > params.range['end'].line
+                then
+                  -- outside of line range
+                  return false
+                end
+
+                if hint_pos.line == params.range.start.line then
+                  -- pos is in the same line as range.start
+                  if hint_pos.line == params.range['end'].line then
+                    -- range.start in the same line as range.end
+                    return params.range.start.character <= hint_pos.character
+                      and hint_pos.character <= params.range['end'].character
+                  end
+                  return hint_pos.character >= params.range.start.character
```

Comment:
please see if `vim.pos` can be used to do position/range comparisons.

(in all parts of this PR, where appropriate)